### PR TITLE
3/video widget

### PR DESCRIPTION
**Pull request by abea**
Resolves https://apostrophecms.atlassian.net/browse/A30U-161, As a user I can edit a video widget.

A few utils and player work done here as well since this is the first core widget with a player. Visually the widget is extremely unopinionated. I could add some styles in the a3-demo to make it nicer.

### CODE REVIEW

File: modules/@apostrophecms/schema/index.js [hunk 1]
```diff
@@ -94,17 +94,6 @@ module.exports = {
       }
     });
 
-    self.addFieldType({
```
Comment (abea) [inline modules/@apostrophecms/schema/index.js:97]: Unrelated, but I think I ran across some error message that led me here.

File: modules/@apostrophecms/schema/ui/apos/components/AposInputArea.vue [hunk 2]
```diff
@@ -19,7 +19,7 @@
 
 <script>
 
-import AposInputMixin from '../mixins/AposInputMixin.js';
+import AposInputMixin from 'Modules/@apostrophecms/schema/mixins/AposInputMixin';
```
Comment (abea) [inline modules/@apostrophecms/schema/ui/apos/components/AposInputArea.vue:22]: I used this syntax to reference this file in the new field component, so it seemed good to switch to this everywhere.

File: modules/@apostrophecms/storybook/template/public/mocks.js [hunk 3]
```diff
@@ -24,7 +24,6 @@
           array: 'AposInputArray',
           object: 'AposInputObject',
           relationship: 'AposInputRelationship',
-          relationshipReverse: 'AposInputRelationshipReverse',
```
Comment (abea) [inline modules/@apostrophecms/storybook/template/public/mocks.js:27]: right? another error message led me here.

File: modules/@apostrophecms/util/ui/public/video.js [hunk 4]
```diff
@@ -1,62 +1,75 @@
-apos.util.widgetPlayers['@apostrophecms/video'] = function(el, data, options) {
+apos.util.widgetPlayers['@apostrophecms/video'] = {
```
Comment (abea) [inline modules/@apostrophecms/util/ui/public/video.js:1]: View this with whitespace changes disabled. Much of it is indented but the same as before.

File: scripts/great-renaming.js [hunk 5]
```diff
@@ -58,7 +58,7 @@ const rename = {
   users: 'user',
   utils: 'util',
   versions: 'version',
-  'video-fields': 'video-field',
+  'video-fields': 'oembed-field',
```
Comment (abea) [inline scripts/great-renaming.js:61]: I'm not sure if this is okay.

File: modules/@apostrophecms/oembed-field/ui/apos/components/AposInputOembed.vue [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+import AposInputMixin from 'Modules/@apostrophecms/schema/mixins/AposInputMixin';
+
+export default {
+  name: 'AposInputOembed',
+  mixins: [ AposInputMixin ],
+  emits: [ 'return' ],
+  data () {
+    return {
+      next: (this.value && this.value.data !== undefined)
+        ? this.value.data : (this.field.def || ''),
```
Comment (boutell) [inline modules/@apostrophecms/oembed-field/ui/apos/components/AposInputOembed.vue:43]: Fields aren't responsible for this, we set up the defaults at the owner's level now. Even in the widget editor.

File: modules/@apostrophecms/oembed-field/ui/apos/components/AposInputOembed.vue [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+          qs: {
+            url: this.next.url
+          }
+        });
+        this.next.title = result.title || '';
+        this.next.thumbnail = result.thumbnail_url || '';
+        this.oembedResult = result;
+
+        if (typeof result.height === 'number' && typeof result.width === 'number') {
+          this.dynamicRatio = (result.height / result.width);
```
Comment (boutell) [inline modules/@apostrophecms/oembed-field/ui/apos/components/AposInputOembed.vue:116]: Nice

File: modules/@apostrophecms/oembed-field/ui/apos/components/AposInputOembed.vue [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+      }
+      return false;
+    },
+    async watchNext () {
+      await this.loadOembed();
+
+      this.validateAndEmit();
+    },
+    async loadOembed () {
+      this.field.disabled = true;
```
Comment (boutell) [inline modules/@apostrophecms/oembed-field/ui/apos/components/AposInputOembed.vue:99]: The field doesn't really belong to us so this is sneaky. But, you do put it back. OK.

File: modules/@apostrophecms/schema/index.js (hunk 1 above)
Comment (boutell) [reply modules/@apostrophecms/schema/index.js:97]: LOL

File: modules/@apostrophecms/schema/ui/apos/components/AposInputArea.vue (hunk 2 above)
Comment (boutell) [reply modules/@apostrophecms/schema/ui/apos/components/AposInputArea.vue:22]: 👍

File: modules/@apostrophecms/storybook/template/public/mocks.js (hunk 3 above)
Comment (boutell) [reply modules/@apostrophecms/storybook/template/public/mocks.js:27]: Good call.

File: modules/@apostrophecms/util/ui/public/utils.js [hunk 9]
```diff
@@ -168,29 +168,16 @@
...
-      if (!player) {
-        return;
-      }
-      player(widget, data, options);
+    for (let i = 0; i < playerList.length; i++) {
+      const playerOpts = players[playerList[i]];
+      const playerEls = (el || document).querySelectorAll(playerOpts.selector);
+
+      playerEls.forEach(function (el) {
+        playerOpts.player(el);
```
Comment (boutell) [inline modules/@apostrophecms/util/ui/public/utils.js:179]: We still want to add data-apos-played and not play them if we have already played them. Yes this is a data attribute of ours on the front end but nobody has to use players if they don't want to.

File: modules/@apostrophecms/util/ui/public/video.js [hunk 10]
```diff
@@ -1,62 +1,75 @@
...
-    apos.util.assign(data.video, {
+    if (!videoUrl || el.getAttribute('data-apos-widget-played')) {
+      return;
+    }
+
+    el.setAttribute('data-apos-widget-played', true);
+
+    queryAndPlay(el, {
+      url: videoUrl,
       neverOpenGraph: 1
```
Comment (boutell) [inline modules/@apostrophecms/util/ui/public/video.js:14]: We should dump our open graph thing, which we never really used effectively in 2.x and just got headaches from. But, that's a ticket for later.

File: modules/@apostrophecms/util/ui/public/video.js [hunk 11]
```diff
@@ -1,62 +1,75 @@
...
+        apos.util.removeClass(el, 'apos-oembed-busy');
+        return play(el, result);
+      });
+    }
 
-  function query(options, callback) {
-    return apos.http.get('/api/v1/@apostrophecms/oembed/query', options, callback);
-  }
+    function query(options, callback) {
+      const body = {
```
Comment (boutell) [inline modules/@apostrophecms/util/ui/public/video.js:33]: It's not a body, that word has a meaning in HTTP. It's an options object.

File: modules/@apostrophecms/video-widget/ui/public/player.js [hunk 12]
```diff
@@ -0,0 +1 @@
+console.info('Hello from the video widget');
```
Comment (boutell) [inline modules/@apostrophecms/video-widget/ui/public/player.js:1]: Cleanup.

File: scripts/great-renaming.js (hunk 5 above)
Comment (boutell) [reply scripts/great-renaming.js:61]: Oh it's fine, this script could be removed at this point, it's historical.

**Review by boutell — CHANGES_REQUESTED** [review]
This is very 💯 , just a few things, most important is that we continue to provide the service of never calling the player twice on the same widget.

File: modules/@apostrophecms/oembed-field/ui/apos/components/AposInputOembed.vue (hunk 6 above)
Comment (abea) [reply modules/@apostrophecms/oembed-field/ui/apos/components/AposInputOembed.vue:43]: That's copied from the input mixin.

File: modules/@apostrophecms/util/ui/public/video.js (hunk 10 above)
Comment (abea) [reply modules/@apostrophecms/util/ui/public/video.js:14]: I added a ticket for this.

**Pull request merged**